### DISCUSSION TITLE
How to generate outputs for all variants

### DISCUSSION THREAD

**Discussion by memen45** [INTI-CMNB/KiBot#646] [Q&A]
It is possible to set the `variant` option for many outputs individually, and it is also possible to set a global `variant`. I use the GitHub Action workflow, and everything is working perfectly, however, I would like to create the renders, gerbers etc. for all the variants available. Since the global `variant` option only accepts a single string, I do not see how to generate for all variants. Any ideas?

**Comment by Cimos** [comment]
Hi memen45,

I can not speak on if there is a command to get kibot to build all the variants, but I can tell you how I do it. 

I have a `variants.yaml` file with each project where we define as many variants a needed.
```yaml
variants:
  - name: 'PN-9999'
    comment: 'PCBA Description'
    type: kibom
    file_id: 'PN:9999'
    variant: 'ADD_USB_C'
```

This file is then included into the `build.kibot.yaml` file that will build everything.
```yaml
import:
  - file: variants.yaml
```

I then have a build script which will use [yq](https://github.com/mikefarah/yq) to query the number of variants in the `variants.yaml` file.
```bash
len=$(yq '.variants | length' "variants,yaml")
```

I'll then loop over building each variant one at a time. Something like this.
```bash
for (( i=0; i<$len; i++ ))
do
    name="$(yq '.variants['${i}'].name' "variants.yaml")"
    kibot --board-file "PN-9999.kicad_pcb" --schematic "PN-9999.kicad_sch" --global-redef "variant=${name}" --out-dir "PCBA_${name} --plot-config "build.kibot.yaml""
   
    # Do some other stuff
done
```

I am curious if there is a better way of doing this. However, it might not work for me as I am doing some other stuff prior to artifact publishing.

Hope this helps

**Reply by set-soft** [reply]
Hi @Cimos !
You can avoid `yq`, just use: `kibot -c CONFIG_FILE --only-names --list-variants` and you'll get a list of available variants.
And while writing this I also added an option to also include sub-PCBs defined in the variants.

**Reply by Cimos** [reply]
@set-soft Thank you for the pointers. Ill update my build script.

P.S Love your work.

**Comment by set-soft** [comment]
Hi @memen45 !

You must do it in separated calls. Like @Cimos suggests.
